src/main/java/hudson/remoting/Channel.java

```diff
@@ -1006,9 +1012,14 @@ public void checkRoles(RoleChecker checker) throws SecurityException {
      * @since 1.299
      */
     public synchronized void join(long timeout) throws InterruptedException {
-        long start = System.currentTimeMillis();
-        while(System.currentTimeMillis()-start<timeout && (inClosed==null || outClosed==null))
-            wait(timeout+start-System.currentTimeMillis());
+        long now = System.currentTimeMillis();
+        long end = now + timeout;
+        while (now < end && (inClosed == null || outClosed == null)) {
+            wait(now - end);
```

Comment:
same as currently happens - instead of waiting for 4 minutes you will wait longer.

This will never be a negative or zero wait due to the "now < end" condition (once I fix the bug).

What will be prevented though is that you will at least wait once (unless inClosed or OutClouse !- null) - whereas before you where never guaranteed to wait.